packages/cannon-worker-api/src/setup.ts

```diff
@@ -0,0 +1,417 @@
+import type { Object3D } from 'three'
+
+import type { AtomicProps, BodyShapeType } from './body'
+
+export type Triplet = [x: number, y: number, z: number]
+export type Quad = [x: number, y: number, z: number, w: number]
+
+export type Broadphase = 'Naive' | 'SAP'
+export type Solver = 'GS' | 'Split'
+export type Buffers = { positions: Float32Array; quaternions: Float32Array }
+export type Refs = { [uuid: string]: Object3D }
+
+export type ConstraintTypes = 'PointToPoint' | 'ConeTwist' | 'Distance' | 'Lock'
+
+export interface ConstraintOptns {
+  collideConnected?: boolean
+  maxForce?: number
+  maxMultiplier?: number
+  wakeUpBodies?: boolean
+}
+
+export interface PointToPointConstraintOpts extends ConstraintOptns {
+  pivotA: Triplet
+  pivotB: Triplet
+}
+
+export interface ConeTwistConstraintOpts extends ConstraintOptns {
+  angle?: number
+  axisA?: Triplet
+  axisB?: Triplet
+  pivotA?: Triplet
+  pivotB?: Triplet
+  twistAngle?: number
+}
+export interface DistanceConstraintOpts extends ConstraintOptns {
+  distance?: number
+}
+
+export interface HingeConstraintOpts extends ConstraintOptns {
+  axisA?: Triplet
+  axisB?: Triplet
+  pivotA?: Triplet
+  pivotB?: Triplet
+}
+
+export type LockConstraintOpts = ConstraintOptns
+
+export interface SpringOptns {
+  damping?: number
+  localAnchorA?: Triplet
+  localAnchorB?: Triplet
+  restLength?: number
+  stiffness?: number
+  worldAnchorA?: Triplet
+  worldAnchorB?: Triplet
+}
+
+export interface WheelInfoOptions {
+  axleLocal?: Triplet
+  chassisConnectionPointLocal?: Triplet
+  customSlidingRotationalSpeed?: number
+  dampingCompression?: number
+  dampingRelaxation?: number
+  directionLocal?: Triplet
+  frictionSlip?: number
+  isFrontWheel?: boolean
+  maxSuspensionForce?: number
+  maxSuspensionTravel?: number
+  radius?: number
+  rollInfluence?: number
+  sideAcceleration?: number
+  suspensionRestLength?: number
+  suspensionStiffness?: number
+  useCustomSlidingRotationalSpeed?: boolean
+}
+
+type WorkerContact = WorkerCollideEvent['data']['contact']
+export type CollideEvent = Omit<WorkerCollideEvent['data'], 'body' | 'target' | 'contact'> & {
+  body: Object3D
+  contact: Omit<WorkerContact, 'bi' | 'bj'> & {
+    bi: Object3D
+    bj: Object3D
+  }
+  target: Object3D
+}
+export type CollideBeginEvent = {
+  body: Object3D
+  op: 'event'
+  target: Object3D
+  type: 'collideBegin'
+}
+export type CollideEndEvent = {
+  body: Object3D
+  op: 'event'
+  target: Object3D
+  type: 'collideEnd'
+}
+export type RayhitEvent = Omit<WorkerRayhitEvent['data'], 'body'> & { body: Object3D | null }
+
+type CannonEvent = CollideBeginEvent | CollideEndEvent | CollideEvent | RayhitEvent
+type CallbackByType<T extends { type: string }> = {
+  [K in T['type']]?: T extends { type: K } ? (e: T) => void : never
+}
+
+// todo - check if used
+export type CannonEvents = { [uuid: string]: Partial<CallbackByType<CannonEvent>> }
```

Comment:
this is used in ProviderContext, can move it there.

Reply:
Ah thanks, will move

Reply:
Moved to `@react-three/cannon` `Provider.tsx` 👍